src/main/java/org/akaza/openclinica/controller/OdmImportController.java

```diff
@@ -29,85 +41,110 @@
-	public String importOdmToOC(@RequestBody PublishDTO publishDTO, HttpServletRequest request)
-			throws Exception {
+    protected final Logger logger = LoggerFactory.getLogger(getClass().getName());
+
+    @RequestMapping(value = "/studies/{studyEnvUuid}/publishJobs", method = RequestMethod.POST)
+    public ResponseEntity<String> importOdmToOC(@PathVariable("studyEnvUuid") String studyEnvUuid, @RequestBody PublishDTO publishDTO, HttpServletRequest request)
+            throws Exception {
 
         String accessToken = (String) request.getSession().getAttribute("accessToken");
+        Study study = studyDao.findByStudyEnvUuid(studyEnvUuid);
+        String studyOid = study.getOc_oid();
+        utilService.setSchemaFromStudyOid(studyOid);
+        study = studyDao.findByStudyEnvUuid(studyEnvUuid);
+        Study site = studyDao.findByOcOID(studyOid);
+        UserAccountBean userAccountBean = utilService.getUserAccountFromRequest(request);
+        UserAccount userAccount = userAccountDao.findById(userAccountBean.getId());
+        String fileName = study.getName();
+
+        //check publish status of study so no 2 processes can occur at the same time
+        if (!jobDetailDao.findStudyIdWithStatus(study.getStudyId(), JobStatus.IN_PROGRESS).isEmpty()) {
+            return new ResponseEntity(ErrorConstants.ERROR_OTHER_PROCESS_IN_PROGRESS, HttpStatus.OK);
+        }
 
-        CompletableFuture<ResponseEntity<Object>> future = CompletableFuture.supplyAsync(() -> {
+        JobDetail jobDetail = userService.persistJobCreated(study, site, userAccount, JobType.PUBLISH_STUDY, fileName);
 
+        CompletableFuture<ResponseEntity<Object>> future = CompletableFuture.supplyAsync(() -> {
+            Map<String, Object> map;
             ODM odm = publishDTO.getOdm();
-            List <Page> pages = publishDTO.getPages();
-            Map<String, Object> map = null;
+            List<Page> pages = publishDTO.getPages();
             try {
                 CoreResources.tenantSchema.set("public");
-                map = (Map<String, Object>) odmImportService.importOdm(odm, pages, publishDTO.getBoardId(), accessToken);
+                map = odmImportService.importOdm(odm, pages, publishDTO.getBoardId(), accessToken);
             } catch (Exception e) {
+                utilService.setSchemaFromStudyOid(studyOid);
+                userService.persistJobFailed(jobDetail, fileName);
                 throw new CompletionException(e);
             }
-            Study study = (Study) map.get("study");
-            Study publicStudy = studyDao.findPublicStudy(study.getOc_oid());
+            Study publicStudy = studyDao.findPublicStudy(studyOid);
             odmImportService.updatePublicStudyPublishedFlag(publicStudy);
             odmImportService.setPublishedVersionsInFM(map, accessToken);
-            return new ResponseEntity<Object>(null, HttpStatus.OK);
+            utilService.setSchemaFromStudyOid(studyOid);
```

Comment:
do we need to switch the schema here again?

Reply:
nope.